### ISSUE TITLE
AsyncGenerator with await incorrectly inferred (different from issue #1912)

### ISSUE THREAD

**Issue by JLessinger** [microsoft/pyright#6948]
This seems closely related to #1912 and related to #2322 and #5153, but it seems to be a distinct bug on 1.1.345.

[See code below] The generator in `run_async_generator()` is incorrectly inferred as a Generator, not AsyncGenerator. In `run_async_generator_with_hint()`, pyright accepts the hint and correctly reports the error in `n2 = next(g_async)`. 

This fails to catch a runtime error that should be statically analyzable. Specifically, the line `n2 = next(g_async)` in `run_async_generator()` is a runtime error, but Pyright accepts it because of the incorrect inference described above.

See trace:

```
Traceback (most recent call last):
  File "gen_bug_repro.py", line 69, in <module>
    asyncio.run(main())
  File "/Users/jonathan/anaconda3/lib/python3.8/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/Users/jonathan/anaconda3/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "gen_bug_repro.py", line 66, in main
    await run_async_generator()
  File "gen_bug_repro.py", line 44, in run_async_generator
    n2 = next(g_async)
TypeError: 'async_generator' object is not an iterator
```

Repro: 

```
import asyncio
from typing import TYPE_CHECKING, AsyncGenerator, Iterator


class SomeIterable:
    def __init__(self):
        self.x = 1

    def __iter__(self) -> Iterator[int]:
        yield self.x


async def foo() -> SomeIterable:
    return SomeIterable()


def bar() -> Iterator[int]:
    yield 2


def run_regular_generator():
    print("regular generator")
    g_regular = (y for y in bar())

    if TYPE_CHECKING:
        reveal_type(g_regular)

    n = next(g_regular)
    print(n)


async def run_async_generator():
    print("async generator")
    g_async = (x + y for y in bar() for x in await foo())

    if TYPE_CHECKING:
        reveal_type(g_async)

    print(f"{type(g_async)}")
    print("anext:")
    n1 = await g_async.__anext__()
    print(n1)
    print("next:")
    n2 = next(g_async)
    print(n2)


async def run_async_generator_with_hint():
    print("async generatorh with hint")
    g_async: AsyncGenerator[int, None] = (x + y for y in bar() for x in await foo())

    if TYPE_CHECKING:
        reveal_type(g_async)

    print(f"{type(g_async)}")
    print("anext:")
    n1 = await g_async.__anext__()
    print(n1)
    print("next:")
    n2 = next(g_async)
    print(n2)


async def main():
    run_regular_generator()
    await run_async_generator()


asyncio.run(main())

```

**Comment by erictraut** [comment]
Mypy also infers the type of `g_async` to be `Generator` rather than `AsyncGenerator`, so if this is a pyright bug, it's also a bug in other type checkers.

Do you happen to know how the runtime decides whether to treat a generator expression as an `AsyncGenerator` or `Generator`? Are all generator expressions within an `async` function assumed to be an `AsyncGenerator`?

Oh, never mind, I just found the relevant text within the [Python documentation](https://docs.python.org/3/reference/expressions.html#generator-expressions):

> If a generator expression contains either async for clauses or [await](https://docs.python.org/3/reference/expressions.html#await) expressions it is called an asynchronous generator expression. An asynchronous generator expression returns a new asynchronous generator object, which is an asynchronous iterator (see [Asynchronous Iterators](https://docs.python.org/3/reference/datamodel.html#async-iterators)).

**Comment by JLessinger** [comment]
> Mypy also infers the type of `g_async` to be `Generator` rather than `AsyncGenerator`, so if this is a pyright bug, it's also a bug in other type checkers.
> 
> 
> 
> Do you happen to know how the runtime decides whether to treat a generator expression as an `AsyncGenerator` or `Generator`? Are all generator expressions within an `async` function assumed to be an `AsyncGenerator`?
> 
> 
> 
> Oh, never mind, I just found the relevant text within the [Python documentation](https://docs.python.org/3/reference/expressions.html#generator-expressions):
> 
> 
> 
> > If a generator expression contains either async for clauses or [await](https://docs.python.org/3/reference/expressions.html#await) expressions it is called an asynchronous generator expression. An asynchronous generator expression returns a new asynchronous generator object, which is an asynchronous iterator (see [Asynchronous Iterators](https://docs.python.org/3/reference/datamodel.html#async-iterators)).

I find the rules confusing too. Thanks for the fix! When will it ship?

**Comment by erictraut** [comment]
> When will it ship?

I typically release a new version of pyright approximately once a week. The last was published three days ago.

**Comment by erictraut** [comment]
This is included in pyright 1.1.346, which I just published. It will also be included in a future release of pylance.
